### ISSUE TITLE
Put together a (pre-GUI) final version of the .yml file

### ISSUE THREAD

**Issue by Niolon** [QCrBox/QCrBox#189]
There are some entries missing in the yml file where we define our commands ([like this one for Olex](https://github.com/QCrBox/QCrBox/blob/dev/services/applications/olex2_linux/config_olex2.yaml)). @maxalbert The cif output is something we need to merge what is new with what is there. Everything else can be vetoed if it is too much work.

Needed:

- [ ] Introduce separate types for cif input / output and general file input/output
- [ ] Handling of cif output
    - [ ] output entries (guaranteed)
    - [ ] output entries (might be there, e.g. extinction, might be tied to input cif entries or input parameters)
    - [ ] invalidated entries (e.g. when you overwrite atomic positions, all quality indicators and _geom are no longer valid and should be deleted)
    - [ ] Optional parameter to select output cif block if program writes multiple
- [ ] Handling of the interactive lifecycle (see #163), defining commands that can do these tasks
    - [ ] prepare
    - [ ] finalise
    - [ ] toparams
    - [ ] redo

Immediately useful.
- [ ] Description fields for commands
- [ ] Description fields for parameters

Make academics happy by assembling citations

- [ ] DOI for program
- [ ] optional DOI for command (in addition to program)

**Comment by Niolon** [comment]
Example
```
#
# Configuration for the application 'olex2'
#
name: "Olex2 (Linux)"
slug: "olex2"
version: "1.5-alpha"
description: "Olex2 is an easy-to-use program containing everything you need to solve, refine and finish small-molecule crystal structures."
url: "https://www.olexsys.org/olex2/"
doi: "10.1107/S0021889808042726"

commands:
  - name: "interactive"
    implemented_as: "interactive"  # other options: "python_callable"
    description: "Start Olex2 in interactive mode, this can be used to solve, refine and finish small-molecule crystal structures."
    parameters:
      - name: "input_cif_path"
        type: "QCrBox.input_cif"
        default_value: None
        description: "Path to the CIF file to open in Olex2."
        required_entry_sets: ["cell_data", "diffraction_data"]
        optional_entry_sets: ["atom_data"]
        merge_su: Yes
        custom_categories: ["iucr", "shelx"]
      - name: "output_cif_path"
        type: "QCrBox.output_cif"
        default_value: None
        description: "Path to save the CIF file to after finishing the interactive session."
        required_entry_sets: ["olex_structure_output", "olex2_only_interactive_output"]
        invalidated_entries: ["_shelx_.*"]
        custom_categories: ["iucr", "shelx"]
        output_block: "0" # select first block from output cif file (the default)
      - name: "parameter_json_path"
        type: "QCrBox.output_file"
        default_value: None
        description: "Path to the JSON file to save the parameters for recreating the interactive result."
      - name: "parameter_folder"
        type: "str"
        default_value: None
        description: "Folder to save files needed as parameters for recreating the interactive result."
    interactive_lifecycle:
      prepare:
        implemented_as: "python_callable"
        module: "configure_olex2"
        function: "prepare__interactive"
        used_basecommand_parameters: ["input_cif_path", "output_cif_path"]
      run:
        implemented_as: "CLI"
        command: "/bin/bash /opt/olex2/start {input_cif_path}"
        used_basecommand_parameters: ["input_cif_path"]
      finalise:
        implemented_as: "python_callable"
        module: "configure_olex2"
        function: "finalise__interactive"
        used_basecommand_parameters: ["input_cif_path", "output_cif_path"]
      toparams:
        implemented_as: "python_callable"
        module: "configure_olex2"
        function: "toparams__interactive"
        used_basecommand_parameters: ["input_cif_path", "output_cif_path", "parameter_json_path", "parameter_folder"]
    non-interactive_equivalent:
      implemented_as: "python_callable"
      module: "configure_olex2"
      function: "redo__interactive"
      parameters:
        - name: "redo_input_cif_path"
          type: "QCrBox.input_cif"
          default_value: None
          description: "Path to the CIF file to recreate the interactive result for."
          required_entry_sets: ["cell_data", "diffraction_data"]
          optional_entry_sets: ["atom_data"]
        - name: "redo_output_cif_path"
          type: "QCrBox.output_cif"
          default_value: None
          description: "Path to save the CIF file to after finishing the interactive session."
          required_entry_sets: ["olex_structure_output", "olex2_only_interactive_output"]
          invalidated_entries: ["_shelx_.*"]
          custom_categories: ["iucr", "shelx"]
          output_block: "0" # select first block from output cif file (the default)
        - name: "parameter_json_path"
          type: "QCrBox.output_file"
          default_value: None
          description: "Path to the JSON file to load the parameters from for recreating the interactive result."
        - name: "parameter_folder"
          type: "QCrBox.output_folder"
          default_value: None
          description: "Folder to load files needed as parameters for recreating the interactive result."
  - name: "refine_iam"
    implemented_as: "CLI"  # other options: "python_callable"
    description: "Refine a CIF file using the IAM method."
    parameters:
      - name: "input_cif_path"
        type: QCrBox.input_cif
        default_value: null
        description: "Path to the CIF file to refine."
        required_entry_sets: ["cell_data", "diffraction_data", "atom_data"]
        merge_su: No
        custom_categories: ["iucr", "shelx"]
      - name: "output_cif_path"
        type: QCrBox.output_cif
        default_value: null
        description: "Path to save the refined CIF file to."
        required_entry_sets: ["olex_structure_output"]
        custom_categories: ["iucr", "shelx"]
        invalidated_entries: ["_shelx_.*"]
      - name: "ls_cycles"
        type: int
        default_value: 5
        required: False
        description: "Number of least squares cycles to run."
      - name: "weight_cycles"
        type: int
        default_value: 5
        required: False
        description: "Number of weight cycles to run. Each weight cycle consists of the full number of least squares cycles."

  - name: "refine_tsc"
    implemented_as: "CLI"  # other options: "python_callable"
    description: "Refine a CIF file using aspheric form factors from a tsc(b) file."
    doi: 10.1039/D0SC05526C
    parameters:
      - name: "input_cif_path"
        type: "QCrBox.input_cif"
        default_value: null
        description: "Path to the CIF file to refine."
        required_entry_sets: ["cell_data", "diffraction_data", "atom_data"]
        merge_su: No
        custom_categories: ["iucr", "shelx"]
      - name: "output_cif_path"
        type: "QCrBox.output_cif"
        default_value: null
        description: "Path to save the refined CIF file to."
        required_entry_sets: ["olex_structure_output"]
        custom_categories: ["iucr", "shelx"]
        invalidated_entries: ["_shelx_.*"]
      - name: "tsc_path"
        type: "QCrBox.input_file"
        default_value: null
        description: "Path to the TSC file to use."
      - name: "ls_cycles"
        type: int
        default_value: 5
        required: False
        description: "Number of least squares cycles to run."
      - name: "weight_cycles"
        type: int
        default_value: 5
        required: False
        description: "Number of weight cycles to run. Each weight cycle consists of the full number of least squares cycles."

  - name: "run_cmds_file"
    implemented_as: "CLI"
    description: "Run a olex console commands read from a text file."
    parameters:
      - name: "input_cif_path"
        type: QCrBox.input_cif
        default_value: null
        description: "Path to the CIF file to run the commands with."
        optional_entry_sets: ["all_olex_optional"]
        custom_categories: ["iucr", "shelx"]
      - name: "output_cif_path"
        type: QCrBox.output_cif
        default_value: null
        description: "Path to save the CIF file to after running the commands. If not given, no output will be read back."
        optional: True
        optional_entry_sets: ["all_olex_optional"]
        custom_categories: ["iucr", "shelx"]
        invalidated_entries: ["_shelx_.*"]
      - name: "cmd_file_path"
        type: "QCrBox.input_file"
        default_value: null
        description: "Path to the file containing the commands to run."

cif_entry_sets:
...
```

**Comment by Niolon** [comment]
The interactive live cycle is missing in this file here

**Comment by Niolon** [comment]
I have added the interactive lifecycle. I do not know how you would define python_callables after refactoring / after the command definition has been moved completely moved to the yml file. So I have made up a placeholder for now

**Comment by Niolon** [comment]
Added invalidated cif entries

**Comment by Niolon** [comment]
Added input output types

**Comment by Niolon** [comment]
- removed unnecessary _cif_ in names now that the context is clear
- More consistent use of io

**Comment by Niolon** [comment]
Added output block to cif_output

**Comment by Niolon** [comment]
- Moved cif information to parameters
 - Changed interactive lifecycle
   - Added run explicitely (Command to start the interactive session)
   - Moved all parameters that are used in any part of the lifecycle to parameters
   - Moved redo out of lifecycle (as it is called separately) and renamed it non-interactive_equivalent

**Comment by Niolon** [comment]
Should also be closed with #203
